pkg/app/app.go

```diff
@@ -789,16 +791,12 @@ func (a *App) visitStates(fileOrDir string, defOpts LoadOpts, converge func(*sta
 		signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
 		go func() {
 			sig := <-sigs
-
-			errs := []error{fmt.Errorf("received [%s] to shutdown ", sig)}
-			_ = context{app: a, st: st, retainValues: defOpts.RetainValuesFiles}.clean(errs)
-			// See http://tldp.org/LDP/abs/html/exitcodes.html
-			switch sig {
-			case syscall.SIGINT:
-				os.Exit(130)
-			case syscall.SIGTERM:
-				os.Exit(143)
-			}
+			func() {
+				CleanWaitGroup.Add(1)
+				defer CleanWaitGroup.Done()
+				errs := []error{fmt.Errorf("received [%s] to shutdown ", sig)}
+				_ = context{app: a, st: st, retainValues: defOpts.RetainValuesFiles}.clean(errs)
+			}()
```

Comment:
I believe adding the task to a wait group "within" a go routine is generally wrong as there can be a race where the `Wait` call can complete before the goroutine starts.

Also, we'd no longer need to add the signal handler in this function as we already set up one in main.go.

That said, how about changing L790-800 to something like the below?

```suggestion
		CleanWaitGroup.Add(1)
		defer CleanWaitGroup.Done()
		go func() {
			CleanWaitGroup.Wait()
			errs := []error{fmt.Errorf("received [%s] to shutdown ", sig)}
			_ = context{app: a, st: st, retainValues: defOpts.RetainValuesFiles}.clean(errs)
		}()
```

Reply:
@mumoshu I will test this in my free time. Thanks very much.